runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/GithubBug_193.java

```diff
@@ -115,6 +147,99 @@ public String toString() {
 		}
 	}
 
+	@Test
+	public void testNestedSchedule() throws Exception {
+		int count = 10000;
+		for (int i = 0; i < count; i++) {
+			twoNestedJobs(i);
+		}
+	}
+
+	private void twoNestedJobs(int i) throws InterruptedException {
+		System.out.println("Starting " + i + " round");
+		int dummyJobsCount = i;
+		final int timeout = 10;
+		for (int j = 0; j < dummyJobsCount; j++) {
+			String name = "dummy_" + j;
+			Job job = Job.create(name, m -> {
+				// nada
+			});
+			if (i % 2 != 0) {
+				job.schedule(i % timeout);
+			} else {
+				job.schedule();
+			}
+			if (i % 2 == 0) {
+				job.join();
+			}
+		}
+
+		Job job1;
+		Job job2;
+		ReentrantLock lock = new ReentrantLock();
+		try {
+			lock.lock();
+
+			String jobName = "job_1_" + i;
+			job1 = new SillyJob(jobName, lock, timeout);
+			jobName = "job_2_" + i;
+			job2 = new SillyJob(jobName, lock, timeout);
+			Object syncObject = new Object();
+			job1.addJobChangeListener(new JobChangeAdapter() {
+				@Override
+				public void running(IJobChangeEvent event) {
+					job2.schedule();
+					synchronized (syncObject) {
+						syncObject.notify();
+					}
+				}
+			});
+
+			synchronized (syncObject) {
+				job1.schedule();
+//				Thread.sleep(10);
+				syncObject.wait();
```

Comment:
Wait without a condition check is likely wrong as this can cause suprious wakeups see:
https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#wait()

Also a notify without condition looks suspicious. You probably want to use a CountDownLatch instead..